### PR TITLE
[Issue-1216] [Issue-1217] add timeout configuration options for go server to cdn.conf

### PR DESCRIPTION
**Pull request by DylanVolz**

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/303/

**Comment by dewrich** [comment]
Added for reference: https://golang.org/pkg/net/#Dialer.Timeout

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/262/

File: traffic_ops/app/conf/cdn.conf [hunk 1]
```diff
@@ -11,6 +11,11 @@
 		workers  => 12,     # Number of workers to run in parallel. Start with a small number.  Increase as need arises.
 	},
 	traffic_ops_golang_port => '443',
+	traffic_ops_golang_proxy_timeout => 60,
```
Comment (dangogh) [inline traffic_ops/app/conf/cdn.conf]: would it be better to create a hash for these?  e.g.
```
traffic_ops_golang => {
   port => '443',
   proxy_timeout => 60,
...
}
```

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 2]
```diff
@@ -96,7 +101,7 @@ func getCDNConf(s string) (Config, error) {
 	}
 	cfg.TOURLStr = "https://127.0.0.1:" + oldPort
 	if cfg.TOURL, err = url.Parse(cfg.TOURLStr); err != nil {
-		return Config{}, fmt.Errorf("Invalid Traffic Ops URL '%v': err", cfg.TOURL, err)
+		return Config{}, fmt.Errorf("Invalid Traffic Ops URL '%v': %s", cfg.TOURL, err)
```
Comment (dangogh) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: I think these are backwards -- cfg.TOURL is %s, err is %v

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 3]
```diff
@@ -109,6 +114,41 @@ func getCDNConf(s string) (Config, error) {
 		return Config{}, err
 	}
 
+	if proxyTimeout, err := getProxyTimeout(obj); err != nil {
+		log.Warnf("failed to get proxy timeout from cdn.conf (%v), using default %v\n", err, DefaultProxyTimeout)
```
Comment (dangogh) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: use %d for timeout value

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 4]
```diff
@@ -109,6 +114,41 @@ func getCDNConf(s string) (Config, error) {
...
+		cfg.ProxyTimeout = DefaultProxyTimeout
+	} else {
+		cfg.ProxyTimeout = proxyTimeout
+	}
+
+	if readTimeout, err := getReadTimeout(obj); err != nil {
+		log.Warnf("failed to get read timeout from cdn.conf (%v), using default %v\n", err, DefaultReadTimeout)
```
Comment (dangogh) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: same as above..   same for each below...

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 5]
```diff
@@ -109,6 +114,41 @@ func getCDNConf(s string) (Config, error) {
...
+
+	if idleTimeout, err := getIdleTimeout(obj); err != nil {
+		log.Warnf("failed to get idle timeout from cdn.conf (%v), using default %v\n", err, DefaultIdleTimeout)
+		cfg.IdleTimeout = DefaultIdleTimeout
+	} else {
+		cfg.IdleTimeout = idleTimeout
+	}
+
 	if dbMaxConns, err := getDBMaxConns(obj); err != nil {
 		log.Warnf("failed to get Max DB Connections from cdn.conf (%v), using default %v\n", err, DefaultMaxDBConnections)
```
Comment (dangogh) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: same here

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 6]
```diff
@@ -136,6 +176,66 @@ func getPort(obj map[string]interface{}) (string, error) {
 	return strconv.Itoa(port), nil
 }
 
+func getProxyTimeout(obj map[string]interface{}) (int, error) {
```
Comment (dangogh) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: a lot of redundant code..  why not pass string index and reduce to a single func?

File: traffic_ops/app/conf/cdn.conf (hunk 1 above)
Comment (dewrich) [reply traffic_ops/app/conf/cdn.conf]: @dangogh My goal is to convert cdn.conf and create a new section for `traffic_ops_golang` that all these configs fall under then the traffic_ops_golang.config can just go away (for the moment)

**Comment by dewrich** [comment]
@DylanVolz did you run the test cases? looks like they are broken now

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/310/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/269/

File: traffic_ops/traffic_ops_golang/perlconfig.go (hunk 2 above)
Comment (DylanVolz) [reply traffic_ops/traffic_ops_golang/perlconfig.go]: TOURL is not a string it is a URL struct and errors implement Stringer.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/313/

File: traffic_ops/traffic_ops_golang/perlconfig.go (hunk 6 above)
Comment (DylanVolz) [reply traffic_ops/traffic_ops_golang/perlconfig.go]: Refactored.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/314/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/272/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/273/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/316/

File: traffic_ops/traffic_ops_golang/traffic_ops_golang.go [hunk 7]
```diff
@@ -61,11 +63,31 @@ func main() {
...
+		Proxy Timeout:       %d
+		Read Timeout:        %d
+		Read Header Timeout: %d
+		Write Timeout:       %d
+		Idle Timeout:        %d
+		Error Log:           %s
+		Warn Log:            %s
+		Info Log:            %s
+		Debug Log:           %s
+		Event Log:           %s`, cfg.HTTPPort, cfg.DBServer, cfg.DBUser, cfg.DBDB, cfg.DBSSL, cfg.MaxDBConnections, cfg.TOURLStr, cfg.Insecure, cfg.CertPath, cfg.KeyPath, cfg.ProxyTimeout, cfg.ReadTimeout, cfg.ReadHeaderTimeout, cfg.WriteTimeout, cfg.IdleTimeout, cfg.LogLocationError, cfg.LogLocationWarning, cfg.LogLocationInfo, cfg.LogLocationDebug, cfg.LogLocationEvent)
```
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/traffic_ops_golang.go]: What about using a nested struct instead, here? It's a lot less code, and people don't have to remember to add things here too.

https://play.golang.org/p/iO2T96zIQI

File: traffic_ops/traffic_ops_golang/traffic_ops_golang.go [hunk 8]
```diff
@@ -61,11 +63,31 @@ func main() {
...
+		Proxy Timeout:       %v
+		Read Timeout:        %v
+		Read Header Timeout: %v
+		Write Timeout:       %v
+		Idle Timeout:        %v
+		Error Log:           %s
+		Warn Log:            %s
+		Info Log:            %s
+		Debug Log:           %s
+		Event Log:           %s`, cfg.HTTPPort, cfg.DBServer, cfg.DBUser, cfg.DBDB, cfg.DBSSL, cfg.MaxDBConnections, cfg.TOURLStr, cfg.Insecure, cfg.CertPath, cfg.KeyPath, time.Duration(cfg.ProxyTimeout) * time.Second, time.Duration(cfg.ReadTimeout) * time.Second, time.Duration(cfg.ReadHeaderTimeout) * time.Second, time.Duration(cfg.WriteTimeout) * time.Second, time.Duration(cfg.IdleTimeout) * time.Second, cfg.LogLocationError, cfg.LogLocationWarning, cfg.LogLocationInfo, cfg.LogLocationDebug, cfg.LogLocationEvent)
```
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/traffic_ops_golang.go]: What about using a nested struct instead, to hide sensitive data? It's a lot less code, and people don't have to remember to add things here too.

```
type ConfigOpen struct {
 Foo int
 Bar string
}
type Config struct {
  ConfigOpen
  Secret string
  Pass string
}
```
```
	cfg := Config {
		ConfigOpen: ConfigOpen{
			Foo: 42,
			Bar: "baz",
		},
		Secret: "walrus",
		Pass: "carpenter",
	}
	fmt.Printf("Config: %+v\n", cfg.ConfigOpen)
```

https://play.golang.org/p/iO2T96zIQI

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/317/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/275/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/318/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/276/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/277/

**Comment by dewrich** [comment]
See PR https://github.com/apache/incubator-trafficcontrol/issues/1217

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/319/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/278/

File: traffic_ops/traffic_ops_golang/traffic_ops_golang.go (hunk 8 above)
Comment (DylanVolz) [reply traffic_ops/traffic_ops_golang/traffic_ops_golang.go]: I like this idea and will see how to make it work smoothly, becuase yes, updating each time a field is added has been a pain point in this PR already.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/325/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/284/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/330/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/289/

File: traffic_ops/app/conf/cdn.conf (hunk 1 above)
Comment (dewrich) [reply traffic_ops/app/conf/cdn.conf]: And creating a hash would mean a change to the `perlhash.go` file

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/338/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/339/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/296/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/297/

**Pull request merged**